shared/src/ledger/masp.rs

```diff
@@ -90,70 +87,37 @@ pub const OUTPUT_NAME: &str = "masp-output.params";
-pub fn load_spend_params() -> (
-    masp_proofs::bellman::groth16::Parameters<Bls12>,
-    masp_proofs::bellman::groth16::PreparedVerifyingKey<Bls12>,
+fn load_pvks() -> (
+    PreparedVerifyingKey<Bls12>,
+    PreparedVerifyingKey<Bls12>,
+    PreparedVerifyingKey<Bls12>,
 ) {
     let params_dir = get_params_dir();
-    let spend_path = params_dir.join(SPEND_NAME);
-    if !spend_path.exists() {
-        #[cfg(feature = "masp_proofs/download-params")]
-        masp_proofs::download_parameters()
-            .expect("MASP parameters not present or downloadable");
-        #[cfg(not(feature = "masp_proofs/download-params"))]
-        panic!("MASP parameters not present or downloadable");
-    }
-    let param_f = File::open(spend_path).unwrap();
-    let params =
-        masp_proofs::bellman::groth16::Parameters::read(&param_f, false)
-            .unwrap();
-    let vk = prepare_verifying_key(&params.vk);
-    (params, vk)
-}
+    let [spend_path, convert_path, output_path] =
+        [SPEND_NAME, CONVERT_NAME, OUTPUT_NAME].map(|p| params_dir.join(p));
 
-/// Load Sapling convert params.
-pub fn load_convert_params() -> (
-    masp_proofs::bellman::groth16::Parameters<Bls12>,
-    masp_proofs::bellman::groth16::PreparedVerifyingKey<Bls12>,
-) {
-    let params_dir = get_params_dir();
-    let spend_path = params_dir.join(CONVERT_NAME);
-    if !spend_path.exists() {
-        #[cfg(feature = "masp_proofs/download-params")]
-        masp_proofs::download_parameters()
-            .expect("MASP parameters not present or downloadable");
-        #[cfg(not(feature = "masp_proofs/download-params"))]
-        panic!("MASP parameters not present or downloadable");
-    }
-    let param_f = File::open(spend_path).unwrap();
-    let params =
-        masp_proofs::bellman::groth16::Parameters::read(&param_f, false)
-            .unwrap();
-    let vk = prepare_verifying_key(&params.vk);
-    (params, vk)
-}
-
-/// Load Sapling output params.
-pub fn load_output_params() -> (
-    masp_proofs::bellman::groth16::Parameters<Bls12>,
-    masp_proofs::bellman::groth16::PreparedVerifyingKey<Bls12>,
-) {
-    let params_dir = get_params_dir();
-    let output_path = params_dir.join(OUTPUT_NAME);
-    if !output_path.exists() {
-        #[cfg(feature = "masp_proofs/download-params")]
-        masp_proofs::download_parameters()
-            .expect("MASP parameters not present or downloadable");
-        #[cfg(not(feature = "masp_proofs/download-params"))]
-        panic!("MASP parameters not present or downloadable");
+    if !spend_path.exists() || !convert_path.exists() || !output_path.exists() {
+        let paths = masp_proofs::download_masp_parameters(None).expect(
+            "MASP parameters were not present, expected the download to \
+             succeed",
+        );
+        if paths.spend != spend_path
+            || paths.convert != convert_path
+            || paths.output != output_path
+        {
+            panic!(
```

Comment:
This needs to be fixed in the future. We shouldn't panic but rather return an error, because otherwise consumers of the SDK can't really handle these failures with error messages.